### ISSUE TITLE
[FSDP] `test_summon_single_param()` is misleading

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#83111]
Code pointer:
https://github.com/pytorch/pytorch/blob/6c60a656b02fbd09661b282bec53940b184db3ca/test/distributed/fsdp/test_fsdp_summon_full_params.py#L269-L288

---

WLOG suppose we have a world size of 2.
- The original parameter (i.e. the linear's weight) is like `[[X]]` where `X` is some initial value and the shape is `[1, 1]`.
- The unpadded unsharded flattened parameter is like `[X]` where the shape is `[1]`.
- The padded unsharded flattened parameter is like `[X, 0]` where the shape is `[2]`. (For larger world sizes, we pad more like `[X, 0, 0, 0]` for world size of 4.)
- After we run on both ranks
    ```
    with torch.no_grad():
        p[0] = self.rank + 2
    ```
    rank 0's local shard is `[2]`, and rank 1's local shard is `[3]`.
- When we run on both ranks
    ```
    with model.summon_full_params(model, writeback=True):
        with torch.no_grad():
           p.copy_(torch.zeros_like(p))
    ```
    `p` is actually the unpadded unsharded flattened parameter on both rank 0 and rank 1. ([code0](https://github.com/pytorch/pytorch/blob/6c60a656b02fbd09661b282bec53940b184db3ca/torch/distributed/fsdp/fully_sharded_data_parallel.py#L2548), [code1](https://github.com/pytorch/pytorch/blob/6c60a656b02fbd09661b282bec53940b184db3ca/torch/distributed/fsdp/fully_sharded_data_parallel.py#L3348)) In other words, if you print `p` in the `summon_full_params()` context, you see `[2]` for both rank 0 and rank 1. This means that the `copy_()` writes to the `0`th element for both ranks. There is not an attempt to zero the `1`st element of the padded unsharded flattened parameter.

We see that this test actually tests whether changes to the padding made before `summon_full_params()` persist after `summon_full_params()`. We could change the `p.copy_(torch.zeros_like(p))` to only run on rank 0, and the test would work just the same.

Personally, I am not sure if FSDP should make any guarantees on persisting writes to the padding. It does not seem like a real use case. 

cc @zhaojuanmao @mrshenli @rohan-varma @ezyang

**Comment by rohan-varma** [comment]
I don't think there's a strong need to test modification to the full flattened unsharded param, as user is not meant to really use this. We should just check that writes to the module params within the context are persisted (this is already technically tested via load_state_dict I believe).

**Comment by mori360** [comment]
We are committed to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
